genie-web/src/main/java/com/netflix/genie/web/data/services/jpa/JpaJobPersistenceServiceImpl.java

```diff
@@ -649,6 +662,12 @@ private void updateJobStatus(
                 jobEntity.setFinished(Instant.now());
             }
         }
+
+        if (currentStatus != newStatus) {
+            synchronized (this.observers) {
+                this.observers.forEach(o -> o.notifyJobStateChange(jobEntity.getUniqueId(), currentStatus, newStatus));
```

Comment:
Ah, i wasn't sure when the change was actually persisted. I think we should indeed try to do it after it is successfully saved. Maybe using this? https://www.baeldung.com/database-auditing-jpa

It's true that the observer is in the critical path, however the current observer actually just puts an event in the genie event bus to be consumed asynchronously by the actual listeners.

Reply:
That article is for entity callbacks... I guess you could do that but then you're bleeding functionality into the entity/model which probably isn't desirable. Maybe wrapping this in an aspect is better so you can have the aspect fire after the method is done?
There's also things like [this](https://dzone.com/articles/transaction-synchronization-and-spring-application) and [this](https://blog.pragmatists.com/spring-events-and-transactions-be-cautious-bdb64cb49a95)

You're assuming here that the current observer is the only observer going forward though. I mean it's fair for our use case but you never know.

Reply:
The downside of all these approaches seems to be that the previous value/entity is not exposed, so all i could emit as event is "state changed, and this is the new state". I cannot say "state changed from X to Y".


There can be multiple observers. But they are responsible for registering themselves, rather than being registered by spring automagically just by existing.